buffer/src/main/java/io/netty/buffer/api/BufferHolder.java

```diff
@@ -0,0 +1,149 @@
+ * in a buffer holder.
+ * <p>
+ * If you just want an object that is a reference to a buffer, then the {@link BufferRef} can be used for that purpose.
+ * If you have an advanced use case where you wish to implement {@link Resource}, and tightly control lifetimes, then
+ * {@link ResourceSupport} can be of help.
+ *
+ * @param <T> The concrete {@link BufferHolder} type.
+ */
+public abstract class BufferHolder<T extends BufferHolder<T>> implements Resource<T> {
+    private static final VarHandle BUF = Statics.findVarHandle(lookup(), BufferHolder.class, "buf", Buffer.class);
+    private Buffer buf;
+
+    /**
+     * Create a new {@link BufferHolder} to hold the given {@linkplain Buffer buffer}.
+     *
+     * @param buf The {@linkplain Buffer buffer} to be held by this holder.
+     */
+    protected BufferHolder(Buffer buf) {
+        this.buf = Objects.requireNonNull(buf, "The buffer cannot be null.");
+    }
+
+    /**
+     * Create a new {@link BufferHolder} to hold the {@linkplain Buffer buffer} received from the given {@link Send}.
+     * <p>
+     * The {@link BufferHolder} will then be holding exclusive ownership of the buffer.
+     *
+     * @param send The {@linkplain Buffer buffer} to be held by this holder.
+     */
+    protected BufferHolder(Send<Buffer> send) {
+        buf = Objects.requireNonNull(send, "The Send-object cannot be null.").receive();
+    }
+
+    @Override
+    public void close() {
+        buf.close();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Send<T> send() {
+        return buf.send().map((Class<T>) getClass(), this::receive);
+    }
+
+    /**
+     * Called when a {@linkplain #send() sent} {@link BufferHolder} is received by the recipient.
+     * The {@link BufferHolder} should return a new concrete instance, that wraps the given {@link Buffer} object.
+     *
+     * @param buf The {@link Buffer} that is {@linkplain Send#receive() received} by the recipient,
+     *           and needs to be wrapped in a new {@link BufferHolder} instance.
+     * @return A new {@linkplain T buffer holder} instance, containing the given {@linkplain Buffer buffer}.
+     */
+    protected abstract T receive(Buffer buf);
+
+    /**
+     * Replace the underlying referenced buffer with the given buffer.
+     * <p>
+     * This method is protected to permit advanced use cases of {@link BufferHolder} sub-class implementations.
+     * <p>
+     * <strong>Note:</strong> This method closes the current buffer,
+     * and takes exclusive ownership of the received buffer.
+     * <p>
+     * The buffer assignment is performed using a plain store.
+     *
+     * @param send The new {@link Buffer} instance that is replacing the currently held buffer.
+     */
+    protected final void replaceBuffer(Send<Buffer> send) {
+        Buffer received = send.receive();
+        buf.close();
```

Comment:
Sanity check: should it be `this.close()` ?

Reply:
No, the `BufferHolder` might assign additional behaviour to its close method. When the buffer being held, we only need to close the replaced buffer. The life time of the holder continues.